### ISSUE TITLE
Region eu-west-2 missing on ansible-2.2 ec2 module

### ISSUE THREAD

**Issue by sjkeerthi** [ansible/ansible#21566]
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
ec2 module

##### ANSIBLE VERSION
2.2

##### SUMMARY
OS: Fedora 24
Ansible: 2.2
Module Version: ec2ansible (0.1.2)

Could not able to create instances via Ansible on London (eu-west-2) Regions

Amazon Elastic Compute Cloud (Amazon EC2)

EU (London)	eu-west-2	ec2.eu-west-2.amazonaws.com	HTTP and HTTPS

ERROR:

TASK [Launch instances] ********************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Region eu-west-2 does not seem to be available for aws module boto.ec2. If the region definitely exists, you may need to upgrade boto or extend with endpoints_path"}

**Comment by jctanner** [comment]
@sjkeerthi please use the issue template in the future.

**Comment by s-hertel** [comment]
I was able to recreate this bug on devel.
```
---
- hosts: localhost
  connection: local
  tasks:
    - name: create ec2 instance with region eu-west-2
      ec2:
        profile: "{{ profile }}"
        #region: us-east-1
        region: eu-west-2
        key_name: "{{ key }}"
        instance_type: t2.micro
        image: "{{ ami }}"
        wait: yes
```
When I try the same playbook with the region us-east-1 (and the region-appropriate key and ami) it works.

**Comment by s-hertel** [comment]
Workaround:
Specify ec2_url: https://ec2.eu-west-2.amazonaws.com and don't use the region option in the playbook.

This error arises using pre boto 2.45. The fix is to upgrade boto!

**Comment by ryansb** [comment]
The lack of eu-west-2 is related to your boto version. On boto v2.45.0 (current on pypi) you can see that the region is available by running:

```
>>> import boto.ec2
>>> boto.ec2.regions()
[RegionInfo:us-east-1, RegionInfo:us-west-1, RegionInfo:us-gov-west-1, RegionInfo:ap-northeast-2, RegionInfo:ap-northeast-1, RegionInfo:sa-east-1, RegionInfo:eu-central-1, RegionInfo:ap-southeast-1, RegionInfo:ca-central-1, RegionInfo:ap-southeast-2, RegionInfo:us-west-2, RegionInfo:us-east-2, RegionInfo:ap-south-1, RegionInfo:cn-north-1, RegionInfo:eu-west-1, RegionInfo:eu-west-2]
```

I'm going to close this issue, please reopen it if upgrading boto doesn't address the problem.

**Comment by sibh8** [comment]
Still an issue at my end. I have upgraded to boto 2.46.1. The python shell for boto module also shows the results for eu-west-2. Still Ansible fails when i use ec2_remote_facts module.

**Comment by s-hertel** [comment]
Hi @sibh8. I have just successfully created, gotten the remote facts for, and terminated an ec2 instance using boto 2.46.1 for eu-west-2. It's unclear what module(s) are failing for you and it appears it may be unrelated to boto 2.46.1. This issue was regarding ec2.py. Please open a new issue with the task(s) that are failing for you and any error messages (use -vvvvv for extra output, please) you are seeing and I'll try to figure out what's going on.

**Comment by wangycc** [comment]
Hi, cn-northwest-1 is not supported in the boto version 2.48.0

```emacs
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Region cn-northwest-1 does not seem to be available for aws module boto.vpc. If the region definitely exists, you may need to upgrade boto or extend with endpoints_path"}
$ sudo pip freeze | grep boto
boto==2.48.0
```

**Comment by s-hertel** [comment]
@wangycc Until that region is added in boto the workaround would be to specify the endpoint using ec2_url like I depicted above.

**Comment by wangycc** [comment]
@s-hertel Hi，I still failed after configuring ec2_ur. Can you help me check it?, thanks

```shell
TASK [vpc : Create VPC] ***************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "region must be specified"}
	to retry, use: --limit @/home/xxxxx/deploy-aws.retry
```
```yaml
- name: "Create VPC"
  ec2_vpc_net:
    name: "{{ vpc.name }}"
    cidr_block: "{{ vpc.cidr_block }}"
    # region: "{{ region }}"
    ec2_url:  https://ec2.cn-northwest-1.amazonaws.com
    tags:
      KubernetesCluster: "{{ k8s_cluster }}"
  register: _vpc
```
version:
```shell
$ ansible --version
ansible 2.3.1.0
  config file = /home/xxxxxxxx/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]

$ pip freeze | grep boto
boto==2.48.0
boto3==1.5.1
botocore==1.8.15
```

**Comment by wangycc** [comment]
```python
    region, ec2_url, aws_connect_params = get_aws_connection_info(module)

    if region:
        try:
            connection = connect_to_aws(boto.vpc, region, **aws_connect_params)
        except (boto.exception.NoAuthHandlerFound, AnsibleAWSError) as e:
            module.fail_json(msg=str(e))
    else:
        module.fail_json(msg="region must be specified")
```
Is this a bug? Code from ansibel stabe2.4: https://github.com/ansible/ansible/blob/stable-2.4/lib/ansible/modules/cloud/amazon/ec2_vpc_net.py

**Comment by s-hertel** [comment]
@wangycc Sorry, notifications for closed issues often slip through the cracks. Using ec2_url as a workaround won't work for ec2_vpc_net on 2.4 because as you noticed, it checks for region and doesn't take the ec2_url parameter. I ported that module to boto3 for 2.5 and removed the check for region and ec2_url is now working correctly for me. If you need this fix now, you can override 2.4's ec2_vpc_net.py by copying the latest module into your local `Library/`.
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_vpc_net.py

**Comment by ghost** [comment]
Still not working on:

```
$ pip list | grep boto
boto            2.49.0
boto3           1.9.27
botocore        1.12.27
```

and

```
$ ansible --version
ansible 2.7.0
```
